include/revng/ABI/CallingConventionTrait.h

```diff
@@ -0,0 +1,1083 @@
+  /// - the third vector register for the `float` argument.
+  /// - the forth GPR register for the `unsigned` argument.
+  ///
+  /// If the ABI is NOT position based, the arguments would be passed in
+  /// - memory for the `Big` argument.
+  /// - the first GPR register for the `signed` argument.
+  /// - the first vector register for the `float` argument.
+  /// - the second GPR register for the `unsigned` argument.
+  ///
+  /// A typical example of a position-based ABI is `Microsoft_x86_64`,
+  /// a non-position-based one - `SystemV_x86_64`.
+  static constexpr bool ArgumentsArePositionBased = false;
+
+  /// States whether an object that needs two GPRs to fit (5-8 bytes on 32-bit
+  /// architectures and 9-16 bytes on 64-bit systems) are only allowed to start
+  /// from a register with an even index.
+  ///
+  /// Here's an example function:
+  /// ```
+  /// void function(uint32_t, uint64_t);
+  /// ```
+  /// On a system with 32-bit wide GPRs, the first argument (`uint32_t`) is
+  /// passed using the first allowed GPR (say `r0`).
+  ///
+  /// The second argument (`uint64_t`) requires two register to fit, so it's
+  /// passed using second and third registers if the ABI allows starting double
+  /// arguments from any register (`r1` and `r2` in this example), or third and
+  /// forth if it only allows starting them from even registers (`r2` and `r3`
+  /// in this example, since `r1` is considered an odd register (the count
+  /// starts from 0, much like C array indexing).
+  ///
+  /// \note: this option is only applicable for non-position based ABIs
+  /// (if `ArgumentsArePositionBased` is `false`).
+  static constexpr bool OnlyStartDoubleArgumentsFromAnEvenRegister = false;
+
+  /// States whether the ABI allows a single object that wouldn't fit into
+  /// a single GRP (9+ bytes on 32-bit systems and 17+ bytes on 64-bit ones)
+  /// to be partially passed in registers with the remainder placed into memory
+  /// if there are not enough registers to fit the entirety of it.
+  ///
+  /// As an example, let's say that there is a big object of type `Big` such
+  /// that `sizeof(Big)` is equal to 16 bytes. On 32-bit system it would mean
+  /// having to use four GPRs (`16 == 4 * 4`) to fit it.
+  ///
+  /// Let's look at an ABI that allocates four registers for passing function
+  /// arguments (`r0-r3`). Then, for a function like
+  /// ```
+  /// void function(uint32_t, Big);
+  /// ```
+  /// the `uint32_t` argument would be passed in the first GPR (`r0`).
+  /// But that would also mean that the remaining three available GPRs are not
+  /// enough to fit the entirety of the `Big` object, meaning it needs to
+  /// either be split between the registers and the memory, or passed using
+  /// the stack. That's exactly what this option states.
+  ///
+  /// \note: this option is only applicable for non-position based ABIs
+  /// (if `ArgumentsArePositionBased` is `false`).
```

Comment:
You really did a great job with explanations, they are now very clear!

Reply:
Ya-a-a-y :D